re-implement filter pushdown for parquet

adriangb opened the pull request:
Very much WIP, please do not review

adriangb commented on datafusion/datasource-parquet/src/file_format.rs:
```diff
@@ -453,29 +451,8 @@ impl FileFormat for ParquetFormat {
-
     fn file_source(&self) -> Arc<dyn FileSource> {
-        Arc::new(ParquetSource::default())
+        todo!() // need access of file schema?
```
This poses an issue.

TLDR is that in order to know if it can absorb a filter as exact `ParquetSource` needs to know not only the filter but also the file schema it's applied to (in particular to get the type of the columns since it can't handle structs).

adriangb commented on datafusion/datasource-parquet/src/source.rs:
```diff
@@ -587,4 +604,40 @@ impl FileSource for ParquetSource {
+        let remaining_description = if config.execution.parquet.pushdown_filters {
+            let mut remaining_filters = fd.filters.clone();
+            for filter in &remaining_filters {
+                if can_expr_be_pushed_down_with_schemas(filter, &conf.file_schema) {
```
This is where we need the file schema

adriangb reviewed (commented):
pointing out current issues to move forward with implementing parquet filter pushdown via the new APIs we've introduced

cc @alamb @berkaysynnada for ideas

alamb commented:
Thanks @adriangb -- I am about to be offline for a week so I will review this when I return

adriangb commented on datafusion/core/src/datasource/listing/table.rs:
```diff
@@ -982,18 +980,6 @@ impl TableProvider for ListingTable {
                     return Ok(TableProviderFilterPushDown::Exact);
                 }
 
-                // if we can't push it down completely with only the filename-based/path-based
-                // column names, then we should check if we can do parquet predicate pushdown
-                let supports_pushdown = self.options.format.supports_filters_pushdown(
-                    &self.file_schema,
-                    &self.table_schema,
-                    &[filter],
-                )?;
-
-                if supports_pushdown == FilePushdownSupport::Supported {
-                    return Ok(TableProviderFilterPushDown::Exact);
-                }
```
The point of this PR is that this moves from being something specialized that `ListingTable` does to anything that works for any TableProvider / they don't need to do anything special! The checks for compatibility also happen all within the parquet data source machinery, instead of leaking implementations via `supports_filters_pushdown`.

adriangb replied:
I think many users of DataFusion (based on our usage, talks I've seen and examples we have) use custom `TableProvider` implementations.

I would keep `supports_filters_pushdown` so that `TableProviders` can do `Exact` pruning of filters, e.g. using partition columns.

alamb replied:
> We can justify implementing other TableProviders for Parquet, but still I cannot understand why we need to degrade the capabilities of our ListingTable. Is't it always better pruning/simplifying things at the higher levels as possible?

I don't think this degrades the capabilities of the current listing table. I think the only implications are for anyone who used a custom `FileFormat` and impleented `supports_filters_pushdown` -- I suspect this is not very common and we can likely avoid consternation by mentioning it in the upgrade guide (see comment below)

adriangb commented on datafusion/core/src/datasource/physical_plan/parquet.rs:
```diff
@@ -148,7 +149,7 @@ mod tests {
 
             let mut source = ParquetSource::default();
             if let Some(predicate) = predicate {
-                source = source.with_predicate(Arc::clone(&file_schema), predicate);
+                source = source.with_predicate(predicate);
```
This seemed like an easy win since I was able to just change this so that the schema is _always_ passed in by the `FileSourceConfigBuilder` instead of only when `with_predicate` is called.
This was necessary becasue `with_predicate` is no longer called to attach a predicate, instaed it happens during an optimization pass so `ParquetSource` neesd to have it available at that point.
I left `with_predicate` in there to avoid churn and in case there is a use case for attaching a predicate directly through the scan instad of a as a FilterExec that later gets pushed into the scan.

adriangb commented on datafusion/datasource-parquet/src/mod.rs:
```diff
@@ -244,7 +242,7 @@ impl ParquetExecBuilder {
             base_config,
             predicate,
-            pruning_predicate: parquet.pruning_predicate,
+            pruning_predicate: None, // for backwards compat since `ParquetExec` is only for backwards compat anyway
```
Open to other suggestions (i.e. removing it). I felt like this minimizes breakage for folks still using `ParquetExec`, who are likely the same folks that want to do the least amount of work possible to upgrade.

alamb replied:
I filed a PR to track this
- https://github.com/apache/datafusion/issues/15950

adriangb commented on datafusion/datasource-parquet/src/row_filter.rs:
```diff
@@ -649,73 +649,45 @@ mod test {
 
     #[test]
     fn nested_data_structures_prevent_pushdown() {
-        let table_schema = get_basic_table_schema();
-
-        let file_schema = Schema::new(vec![Field::new(
-            "list_col",
-            DataType::Struct(Fields::empty()),
-            true,
-        )]);
```
This test was wrong! It wanted to test that `list_col` prevents pushdown because it's a nested type. Instead it was prevented because `list_col` is not in the table / schema!

adriangb commented on datafusion/datasource-parquet/src/source.rs:
```diff
@@ -559,25 +549,8 @@ impl FileSource for ParquetSource {
                     .predicate()
                     .map(|p| format!(", predicate={p}"))
                     .unwrap_or_default();
-                let pruning_predicate_string = self
-                    .pruning_predicate
-                    .as_ref()
-                    .map(|pre| {
-                        let mut guarantees = pre
-                            .literal_guarantees()
-                            .iter()
-                            .map(|item| format!("{}", item))
-                            .collect_vec();
-                        guarantees.sort();
-                        format!(
-                            ", pruning_predicate={}, required_guarantees=[{}]",
-                            pre.predicate_expr(),
-                            guarantees.join(", ")
-                        )
-                    })
-                    .unwrap_or_default();
```
In https://github.com/apache/datafusion/pull/15561#pullrequestreview-2741008179 Andrew asked me to keep this, but now since the schema isn't even being passed in to `with_predicate` it's going to be hard to keep these. I suggest we just accept that they won't be present in the physical plans. If that's not okay what I could do is generate them on the fly in `fmt_extra` or generate them if `with_predicate` is called with a schema or `with_schema` is called with a predicate. But I'd like to avoid that unless someone thinks is worth it or has another suggestion.

alamb replied:
I think it is important to keep these in the physical plans -- in particular what I think is important is to be able to check via the explain plan if pruning is happening by looking at the explain plan

adriangb replied:
Hmm okay. I'll see if I can make it happen...

adriangb commented on datafusion/datasource-parquet/src/source.rs:
```diff
@@ -587,4 +560,49 @@ impl FileSource for ParquetSource {
         }
     }
+
+    fn try_pushdown_filters(
```
cc @berkaysynnada for this implementation

adriangb commented on datafusion/datasource/src/file_format.rs:
```diff
@@ -109,37 +108,10 @@ pub trait FileFormat: Send + Sync + fmt::Debug {
         not_impl_err!("Writer not implemented for this format")
     }
 
-    /// Check if the specified file format has support for pushing down the provided filters within
-    /// the given schemas. Added initially to support the Parquet file format's ability to do this.
-    fn supports_filters_pushdown(
```
Binning specialized code that was also leaking parquet stuff through DataSource and into TableProvider 😄

alamb replied:
yes I agree 

Since [`FileFormat`](https://docs.rs/datafusion/latest/datafusion/datasource/file_format/trait.FileFormat.html) is a pub trait, this is technically a breaking API change, but I do think it was a parquet specific optimization

I recommend we mark this PR as an API change and add a note to the upgrade guide https://github.com/apache/datafusion/blob/main/docs/source/library-user-guide/upgrading.md

I think it should basically say if you implemented `FileFormat` (which probably no one did) and ListingTable you will have to implement the newly added `ExecutionPlan::try_pushdown_filter` method into your execution plan directly if you want the filters to be pushed down

adriangb commented on datafusion/datasource/src/file_format.rs:
```diff
@@ -109,37 +108,10 @@ pub trait FileFormat: Send + Sync + fmt::Debug {
-/// An enum to distinguish between different states when determining if certain filters can be
-/// pushed down to file scanning
-#[derive(Debug, PartialEq)]
-pub enum FilePushdownSupport {
```
Another one of these enums!

adriangb commented on datafusion/sqllogictest/test_files/parquet_filter_pushdown.slt:
```diff
@@ -81,11 +81,15 @@ EXPLAIN select a from t_pushdown where b > 2 ORDER BY a;
 logical_plan
 01)Sort: t_pushdown.a ASC NULLS LAST
-02)--TableScan: t_pushdown projection=[a], full_filters=[t_pushdown.b > Int32(2)]
+02)--Projection: t_pushdown.a
+03)----Filter: t_pushdown.b > Int32(2)
+04)------TableScan: t_pushdown projection=[a, b], partial_filters=[t_pushdown.b > Int32(2)]
```
This is because the pushdown no longer happens at the logical level - it happens at the physical level. This makes sense, in part because the checks for suitability of pushdown are better at the physical level (there may be reasons to reject a pushdown at the physical level that are not present at the logical level, e.g. partition columns or encodings).

alamb replied:
I think it makes sense and is ok that the logical plans show the filter not pushed down

adriangb commented on datafusion/sqllogictest/test_files/parquet_filter_pushdown.slt:
```diff
@@ -81,11 +81,15 @@ EXPLAIN select a from t_pushdown where b > 2 ORDER BY a;
+04)------TableScan: t_pushdown projection=[a, b], partial_filters=[t_pushdown.b > Int32(2)]
 physical_plan
 01)SortPreservingMergeExec: [a@0 ASC NULLS LAST]
 02)--SortExec: expr=[a@0 ASC NULLS LAST], preserve_partitioning=[true]
-03)----DataSourceExec: file_groups={2 groups: [[WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/parquet_filter_pushdown/parquet_table/1.parquet], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/parquet_filter_pushdown/parquet_table/2.parquet]]}, projection=[a], file_type=parquet, predicate=b@1 > 2, pruning_predicate=b_null_count@1 != row_count@2 AND b_max@0 > 2, required_guar ...
+03)----CoalesceBatchesExec: target_batch_size=8192
+04)------RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=2
+05)--------DataSourceExec: file_groups={2 groups: [[WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/parquet_filter_pushdown/parquet_table/1.parquet], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/parquet_filter_pushdown/parquet_table/2.parquet]]}, projection=[a], file_type=parquet, predicate=b@1 > 2 AND b@1 > 2
```
@berkaysynnada any idea why we have extra `CoalesceBatchesExec` and `RepartitionExec` now?

berkaysynnada replied:
I've a guess but not proved: CoalesceBatchesExec comes because of RepartitionExec, and RepartitionExec is inserted to satisfy partition count, which is 4. That's required by FilterExec now (which was pushed down at the logical level before), but that FilterExec is pushed down later after EnforceDistribution.

So, this makes me think about the correct order of physical rules. PushdownFilter should probably work before distribution&order satisfiers.  But that could also bring some issues, I'm not sure.

adriangb replied:
> PushdownFilter should probably work before distribution&order satisfiers
That makes sense to me. It does more "invasive" re-arranging of plans than those do.

alamb replied:
I agree it is important to remove the coalesce / repartition

Maybe we can make a separate PR to move the filter pushdown code earlier in the physical planning

An alternate could be to update the filter pushdown optimizer pass somehow to remove these -- but I think it would be cleaner / easier to understand if they were never added in the first place

adriangb replied:
This would also be fixed by https://github.com/apache/datafusion/pull/15801

adriangb replied:
I opened https://github.com/apache/datafusion/issues/15938

adriangb commented on datafusion/sqllogictest/test_files/aggregate.slt:
```diff
@@ -5006,7 +5006,7 @@ SELECT column5, avg(column1) FROM d GROUP BY column5;
 query I??
 SELECT column5, column1, avg(column1) OVER (PARTITION BY column5 ORDER BY column1 ROWS BETWEEN 1 PRECEDING AND CURRENT ROW) as window_avg 
-FROM d WHERE column1 IS NOT NULL;
+FROM d WHERE column1 IS NOT NULL ORDER BY column5 DESC;
```
Without this in place the order of the rows changed. I'm not sure why, but I also don't think anything guarantees it without an `ORDER BY` clause so it seemed sensible to add one. A second opinion would be valuable.

berkaysynnada replied:
Plan doesn't change, but why results changed 🤔

adriangb replied:
Yeah I'm a bit confused as well. It's just the order of the results that changed. I feel like there must be something trivial causing this that we can revert.

jayzhan211 replied:
You probably need to rebase, I find the result in main branch is the same as your

adriangb replied:
Thank you indeed that fixed it 😄

adriangb commented:
> Thanks @adriangb -- I am about to be offline for a week so I will review this when I return

Enjoy your vacation! I think you'll like this diff:

<img width="144" alt="image" src="https://github.com/user-attachments/assets/d9692934-910b-4974-95dd-fb9ed7beaca3" />

berkaysynnada commented on datafusion/datasource-parquet/src/source.rs:
```diff
@@ -253,18 +251,18 @@ use object_store::ObjectStore;
+    pub(crate) file_schema: Option<SchemaRef>,
     /// Optional predicate for row filtering during parquet scan
     pub(crate) predicate: Option<Arc<dyn PhysicalExpr>>,
-    /// Optional predicate for pruning row groups (derived from `predicate`)
```
good to see these are unifying

berkaysynnada commented on datafusion/datasource-parquet/src/source.rs:
```diff
@@ -253,18 +251,18 @@ use object_store::ObjectStore;
+    /// The schema of the file.
+    /// In particular, this is the schema of the table without partition columns,
+    /// *not* the physical schema of the file.
+    pub(crate) file_schema: Option<SchemaRef>,
```
There is also another schema in FileScanConfig. Are they both reflects the file schema, not physical schema? and can we somehow unify them?

adriangb replied:
This is the same schema that `FileScanConfig` passes into `ParquetSource`

berkaysynnada reviewed (commented):
Thank you @adriangb. I couldn't provide much design suggestions, since I cannot  fully understand the need of these changes. If you provide more background information, I can help more maybe. 

It seems there are some critical planning changes here, and it's better getting approvals by more people for this PR.

adriangb commented on datafusion/datasource-parquet/src/source.rs:
```diff
@@ -589,4 +559,49 @@ impl FileSource for ParquetSource {
+        };
+        let config_pushdown_enabled = config.execution.parquet.pushdown_filters;
+        let table_pushdown_enabled = self.pushdown_filters();
+        if table_pushdown_enabled || config_pushdown_enabled {
```
The current behavior is not documented anywhere, I tried to match the existing tests: https://github.com/apache/datafusion/blob/9730404028a91a7fe875ea3f88bafdbcb305ae6c/datafusion/sqllogictest/test_files/parquet_filter_pushdown.slt#L44-L88

adriangb commented on datafusion/datasource-parquet/src/source.rs:
```diff
@@ -589,4 +559,49 @@ impl FileSource for ParquetSource {
+        fd: FilterDescription,
+        config: &datafusion_common::config::ConfigOptions,
+    ) -> datafusion_common::Result<FilterPushdownResult<Arc<dyn FileSource>>> {
+        let Some(file_schema) = self.file_schema.clone() else {
```
I think they always end up with a schema now, but the current APIs don't require it via the constructor and instead it gets passed in via `FileScanConfigBuilder`. I tried piping it into the constructor but makes things difficult, there's APIs that rely on `ParquetSource::default()` and such. So TLDR is it's a bit gross but this is the least chrun way to do it and we can always come back later and clean the rest up.

alamb replied:
Let's file that follow on ticket

adriangb replied:
https://github.com/apache/datafusion/issues/15952

adriangb commented on datafusion/datasource-parquet/src/source.rs:
```diff
@@ -589,4 +559,49 @@ impl FileSource for ParquetSource {
+        let config_pushdown_enabled = config.execution.parquet.pushdown_filters;
+        let table_pushdown_enabled = self.pushdown_filters();
+        if table_pushdown_enabled || config_pushdown_enabled {
+            let mut conf = self.clone();
```
It's the pattern used everywhere else but I'll change it

adriangb replied:
8c1b98cf8

adriangb commented:
https://github.com/apache/datafusion/issues/15812 surfaced another reason why building the predicates from the files schemas is necessary

adriangb commented on datafusion/datasource-parquet/src/source.rs:
```diff
@@ -589,4 +559,49 @@ impl FileSource for ParquetSource {
+            let mut conf = self.clone();
+            let mut allowed_filters = vec![];
+            let mut remaining_filters = vec![];
+            for filter in &fd.filters {
```
f2bec87e2

berkaysynnada reviewed (commented):
There are good things here, but the main change doesn't seem correct to me. Why are we reducing the capabilities of logical optimizations? I think these planning changes will harm some people. How does it block the dynamic filtering approach?

adriangb commented:
I'm not really sure how this degrades anything. The end result is the same, users won't see any difference.

What ListingTable does currently is misguided and wrong since it is not really at the logical level as you say, instead it pierces the logical / physical separation (see how it converts Expr to PhysicalExpr, etc). It even produces bugs (I believe the pushdown of struct fields may currently be broken, or at least the implementation is confusing and the test is completely wrong).

I think there is pushdown that makes sense at the logical level, namely partition pruning. And I left that for TableProvider to continue to do. But the pruning that relies on a PhysicalExpr seems to me like it should be happening at the physical layer not the logical.

We might be able to leave the stuff in TableProvider in place but we'll be dealing with duplication and confusing methods on DataSource, which is already a complex bit of code. When I first tried to implement it this way I ran into cases with duplicate pushdown and other confusing scenarios. Probably it could have been resolved but I felt like why make one of the most complex bits in DataFusion even more complex instead of simplifying it where possible.

berkaysynnada commented:
> I'm not really sure how this degrades anything. The end result is the same, users won't see any difference.

Logical planning results are changing. We are also using DF end-to-end, but there could always be people only relying on logical plans of DF. 

> We might be able to leave the stuff in TableProvider in place but we'll be dealing with duplication and confusing methods on DataSource, which is already a complex bit of code. When I first tried to implement it this way I ran into cases with duplicate pushdown and other confusing scenarios. Probably it could have been resolved but I felt like why make one of the most complex bits in DataFusion even more complex instead of simplifying it where possible.

I'm also challenging to decide to be which side because of that complexity :D

I will take a look to other parts in this PR, and try to find a solution for the points like https://github.com/apache/datafusion/pull/15769/files#r2051612926. Maybe there isn't something wrong at all as you said, there is no harm to double check

alamb commented on datafusion/sqllogictest/test_files/push_down_filter.slt:
```diff
@@ -218,43 +219,57 @@ LOCATION 'test_files/scratch/push_down_filter/t.parquet';
-logical_plan TableScan: t projection=[a], full_filters=[t.a != Int32(100)]
+logical_plan
+01)Filter: t.a != Int32(100)
+02)--TableScan: t projection=[a], partial_filters=[t.a != Int32(100)]
```
I recommend we change these tests to show the physical plans (not the logical plans) as that would more accurately show the pushdown happening. Maybe also something we could do as a separate PR

adriangb replied:
@alamb do you know why these don't display the phyiscal plan already? Is something parsing them out?

berkaysynnada replied:
There is a statement in 21st line in this file:
`set datafusion.explain.logical_plan_only = true;`

There is also counter-part of that: `set datafusion.explain.physical_plan_only = true`

You can set the display type of explain queries with these.

alamb reviewed (approved):
Ok, I went through this PR and TLDR is I think it is an improvement. Thank you very much @adriangb 

I left several suggestions on how to improve it / the tests, but I also think we could do that as a follow on PR. 

I think this is a really nice step forward -- and while it is taking a long time I am confident it will be worth it in the end

berkaysynnada commented:
@adriangb do you have time to address the last suggestions? I understand the mistake here, and I think we should take this in asap

adriangb commented:
> @adriangb do you have time to address the last suggestions? I understand the mistake here, and I think we should take this in asap

I am going to try to address the last round of review later today on a flight. In particular:
- https://github.com/apache/datafusion/pull/15769#discussion_r2070806936: change the relevant tests to show the physical plan
- https://github.com/apache/datafusion/pull/15769#discussion_r2070258342: try to show _a_ predicate in explain plans (it will not exactly match the final predicate though since the latter will be generated based on the physical file schema)

Is there anything I'm missing? Is this what you meant by `the mistake`?

berkaysynnada commented:
> Is there anything I'm missing? Is this what you meant by `the mistake`?

I meant the need of this PR.

> Is there anything I'm missing?
https://github.com/apache/datafusion/pull/15769#discussion_r2070398720 
a ticket would be good IMO as well
https://github.com/apache/datafusion/pull/15769#discussion_r2070804563
do you wanna give it a try to change the orders of rules?

adriangb commented:
I updated the order of the pushdown rules in this PR, it worked to get rid of the extra nodes.

I've also added the upgrade guide and the pushdown preview is being shown in the physical plans.

@alamb I think the only point missing is https://github.com/apache/datafusion/pull/15769#discussion_r2070806936 which I need a big of guidance on

berkaysynnada reviewed (approved):
Once the conflicts are resolved, I'll merge this: Thanks @adriangb

berkaysynnada commented:
I think you can disable logical plans. That's what @alamb suggests, if I don't get it wrong

berkaysynnada commented:
I think we should also inform people who might be possibly effected by these logical plan changes

@andygrove

adriangb commented:
Okay folks this is rebased and ready for another round of review

alamb commented on datafusion/core/tests/sql/path_partition.rs:
```diff
@@ -57,55 +53,6 @@ use object_store::{
 use url::Url;
 
-#[tokio::test]
-async fn parquet_partition_pruning_filter() -> Result<()> {
```
can you remind me again why this test was removed ?

adriangb replied:
It was relying on `ListingTable` to do the pushdown without the optimizer pass.
I felt that it was better suited for an `.slt` test that is fully wired up, so I essentially moved it: https://github.com/apache/datafusion/pull/15769/files#r2074193501

alamb commented on datafusion/sqllogictest/test_files/push_down_filter.slt:
```diff
@@ -18,7 +18,7 @@
 
 statement ok
-set datafusion.explain.logical_plan_only = true;
+set datafusion.explain.physical_plan_only = true;
```
💯

alamb commented on datafusion/sqllogictest/test_files/push_down_filter.slt:
```diff
@@ -179,9 +190,9 @@ LOCATION 'test_files/scratch/parquet/test_filter_with_limit/';
-02)--TableScan: test_filter_with_limit projection=[part_key, value], full_filters=[test_filter_with_limit.value = Int32(2)], fetch=1
+physical_plan
+01)CoalescePartitionsExec: fetch=1
+02)--DataSourceExec: file_groups={3 groups: [[WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/parquet/test_filter_with_limit/part-0.parquet], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/parquet/test_filter_with_limit/part-1.parquet], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/parquet/test_filter_with_limit/part-2.parquet]]}, projection=[part_key, value], l ...
```
it is good to see this the predicate pushed down

alamb commented on docs/source/library-user-guide/upgrading.md:
```diff
@@ -46,6 +46,16 @@ schema. To upgrade structs which implement `PhysicalExpr` you need to implement
 the `return_field` function. There are numerous examples in the `physical-expr`
 crate.
 
+### `FileFormat::supports_filters_pushdown` replaced with `FileSource::try_pushdown_filters`
```
👍

alamb reviewed (approved):
Thank you @adriangb  -- I think it looks great. I'll plan to merge it tomorrow unless anyone else would like a chance to review

alamb commented:
It looks like there are a few doc errors and doc exaple errors

adriangb commented on datafusion/sqllogictest/test_files/parquet_filter_pushdown.slt:
```diff
@@ -221,5 +226,23 @@ physical_plan DataSourceExec: file_groups={2 groups: [[WORKSPACE_ROOT/datafusion
+02)--TableScan: t_pushdown projection=[val, part], partial_filters=[t_pushdown.val != Utf8("c")]
 physical_plan DataSourceExec: file_groups={3 groups: [[WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/parquet_filter_pushdown/parquet_part_test/part=a/file.parquet], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/parquet_filter_pushdown/parquet_part_test/part=b/file.parquet], [WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/parquet_filter_pushdown/parquet_part_tes ...
+
+# If we have a mix of filters:
+# - The partition filters get evaluated during planning
+# - The mixed filters end up in a FilterExec
+# - The file filters get pushed down into the scan
+query TT
+EXPLAIN select * from t_pushdown where val != 'd' AND val != 'c' AND part = 'a' AND part != val;
+----
+logical_plan
+01)Filter: t_pushdown.val != Utf8("d") AND t_pushdown.val != Utf8("c") AND t_pushdown.val != t_pushdown.part
+02)--TableScan: t_pushdown projection=[val, part], full_filters=[t_pushdown.part = Utf8("a")], partial_filters=[t_pushdown.val != Utf8("d"), t_pushdown.val != Utf8("c"), t_pushdown.val != t_pushdown.part]
+physical_plan
+01)CoalesceBatchesExec: target_batch_size=8192
+02)--FilterExec: val@0 != part@1
+03)----RepartitionExec: partitioning=RoundRobinBatch(4), input_partitions=1
+04)------DataSourceExec: file_groups={1 group: [[WORKSPACE_ROOT/datafusion/sqllogictest/test_files/scratch/parquet_filter_pushdown/parquet_part_test/part=a/file.parquet]]}, projection=[val, part], file_type=parquet, predicate=val@0 != d AND val@0 != c, pruning_predicate=val_null_count@2 != row_count@3 AND (val_min@0 != d OR d != val_max@1) AND val_null_count@2 != row_count@3 AND (val_min@0 != c O ...
```
Replaces deleted test

adriangb commented on datafusion/physical-optimizer/src/optimizer.rs:
```diff
@@ -95,6 +95,10 @@ impl PhysicalOptimizer {
             // as that rule may inject other operations in between the different AggregateExecs.
             // Applying the rule early means only directly-connected AggregateExecs must be examined.
             Arc::new(LimitedDistinctAggregation::new()),
+            // The FilterPushdown rule tries to push down filters as far as it can.
+            // For example, it will push down filtering from a `FilterExec` to
+            // a `DataSourceExec`, or from a `TopK`'s current state to a `DataSourceExec`.
+            Arc::new(FilterPushdown::new()),
```
I guess the only logical solution is to place `FilterPushdown` at the end and remove any unnecessary repartitions?

berkaysynnada replied:
If this version is okay in this PR, I can take a look at that in #15770. I've some ideas, like making FilterPushdown smarter to remove any unnecessary patterns behind

berkaysynnada commented on datafusion/physical-plan/src/filter_pushdown.rs:
```diff
@@ -98,6 +116,13 @@ impl PredicateSupports {
     }
 
+    /// Are all filters marked as [`PredicateSupport::Supported`]?
+    pub fn is_exact(&self) -> bool {
```
can we find a better name for this? and it's not used, maybe removed until it's needed?

adriangb replied:
Already removed in 15443d5

alamb commented:
WOOHOO!

xudong963 commented on datafusion/datasource/src/file_format.rs:
```diff
@@ -94,7 +93,6 @@ pub trait FileFormat: Send + Sync + fmt::Debug {
         &self,
         state: &dyn Session,
         conf: FileScanConfig,
-        filters: Option<&Arc<dyn PhysicalExpr>>,
```
This is definitely a breaking API change, we depend on the interface intensively 😢

xudong963 replied:
Thank you @alamb, I'll add this to my  TODO list.

adriangb replied:
Agreed. Sorry for the breakage. Do the new APIs get the job done for you, or is there something now missing?

xudong963 commented:
There is a bug related to the PR: https://github.com/apache/datafusion/issues/16563
